### PR TITLE
Bb displaying post details

### LINKED ISSUE #43: View Post Details
As a reader, I would like to see the content of a Post so I can read it.

**Given** a user is viewing a list of Posts  
**When** they select a post to read  
**Then** they should be directed to a Post Detail page that shows the Post Details.

Post Details include:

* Title
* Header image (if exists)
* Content
* Publication date (MM/DD/YYYY)
* Author's Display Name

### PR DESCRIPTION
**Pull request by Borba02**
#### Changes Made
1. Added file GetPostById method to API
2. Added `PostDetails.js` component to client
3. Modified file `ApplicationViews.js` to include routes for posts details
​
#### Steps to Review
1. Checkout this branch locally.
    ```
    git fetch --all
    git checkout bb-DisplayingPostDetails
    ```
7. Open a new Terminal tab (⌘T) and navigate to the server directory.
8. Test app functionality.
    > When a logged in user clicks on the Posts link they are taken to a list of posts and each entry in the list has author name, title, content, and a details link. When the user clicks the details link they should be directed to a view displaying more details about that post
9. View code file.
    > Confirm file modifications are present as indicated above.
    > Confirm no unused code or extraneous comments exist.

Closes #43

### CODE REVIEW

**Review by JacobF94 — APPROVED** [review]
Tested functionality, all working!

File: Tabloid/Repositories/PostRepository.cs [hunk 1]
```diff
@@ -70,6 +70,66 @@ public List<Post> GetAll()
...
+                        {
+                            post = new Post()
+                            {
+                                Id = id,
+                                Title = DbUtils.GetString(reader, "Title"),
+                                Content = DbUtils.GetString(reader, "Content"),
+                                ImageLocation = DbUtils.GetString(reader, "ImageLocation"),
+                                CreateDateTime = DbUtils.GetDateTime(reader, "CreateDateTime"),
+                                PublishDateTime = DbUtils.GetDateTime(reader, "PublishDateTime"),
+                                IsApproved = reader.GetBoolean(reader.GetOrdinal("IsApproved")),
```
Comment (heymonicakay) [inline Tabloid/Repositories/PostRepository.cs]: If there's no DbUtils method to support getting a bool, maybe you want to add one?

File: Tabloid/Repositories/PostRepository.cs [hunk 2]
```diff
@@ -70,6 +70,66 @@ public List<Post> GetAll()
...
             }
         }
+        public Post GetPostById(int id)
+        {
+            using (var conn = Connection)
+            {
+                conn.Open();
+                using (var cmd = conn.CreateCommand())
+                {
+                    cmd.CommandText = @"SELECT p.Title, p.Content, p.ImageLocation, p.CreateDateTime, p.PublishDateTime, p.IsApproved, p.CategoryId, p.UserProfileId,
```
Comment (heymonicakay) [inline Tabloid/Repositories/PostRepository.cs:80]: Thanks for using consistent casing here!

File: Tabloid/client/src/components/ApplicationViews.js [hunk 3]
```diff
@@ -19,7 +20,10 @@ export default function ApplicationViews({ isLoggedIn }) {
           <CategoryList />
         </Route>
         <Route path="/posts" exact>
-          {isLoggedIn ? <PostList /> : <Redirect to="/login" />}
+          <PostList />
+        </Route>
+        <Route path="/posts/:postId" exact>
+          <PostDetails />
```
Comment (heymonicakay) [inline Tabloid/client/src/components/ApplicationViews.js]: Should the user be redirected to the login view if they are not logged in?

File: Tabloid/client/src/components/ApplicationViews.js [hunk 4]
```diff
@@ -19,7 +20,10 @@ export default function ApplicationViews({ isLoggedIn }) {
           <CategoryList />
         </Route>
         <Route path="/posts" exact>
-          {isLoggedIn ? <PostList /> : <Redirect to="/login" />}
+          <PostList />
+        </Route>
+        <Route path="/posts/:postId" exact>
```
Comment (heymonicakay) [inline Tabloid/client/src/components/ApplicationViews.js]: 1 - Does this route path need to exact?
2 - Should this route param be an integer? if so, you can add some regex to make sure it is an integer.. see client side chapters on dynamic routing for examples!

File: Tabloid/client/src/components/posts/PostDetails.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import React from "react";
+import { Card, CardBody } from "reactstrap";
+import { useState, useEffect } from "react";
+import { getPostById } from "../../modules/postManager";
+import { useParams } from "react-router-dom";
+
+const PostDetails = () => {
+  const [post, setPosts] = useState({});
```
Comment (heymonicakay) [inline Tabloid/client/src/components/posts/PostDetails.js]: The naming convention here between the state variable and it's setter is inconsistent (singular v plural).. is this intentional? If so, what is the reason?

File: Tabloid/client/src/components/posts/PostList.js [hunk 6]
```diff
@@ -15,10 +17,17 @@ const PostList = () => {
...
-      <div className="row">
-        {posts.map((post) => (
-          <Post post={post} key={post.id} />
-        ))}
+      <div className="row justify-content-center">
+        <ListGroup>
+          {posts.map((post) => {
+            return (
+              <ListGroupItem key={post.id}>
+                <Post post={post} key={post.id} />
```
Comment (heymonicakay) [inline Tabloid/client/src/components/posts/PostList.js]: If there is a key attribute on the parent ListGroupItem component, is a key attribute also required on the Post component?

**Review by heymonicakay — CHANGES_REQUESTED** [review]
Please see my comments.

File: Tabloid/client/src/components/posts/PostList.js (hunk 6 above)
Comment (Borba02) [reply Tabloid/client/src/components/posts/PostList.js]: I've consolidated the key attribute to just the ListGroupItem and I also added an alt attribute to the post header image's html img tag. I was getting a compiler warning about not having done so.

File: Tabloid/client/src/components/posts/PostDetails.js (hunk 5 above)
Comment (Borba02) [reply Tabloid/client/src/components/posts/PostDetails.js]: I have an unhealthy relationship with my 's' key. I tend to use it when I don't need to and forget when I do.

Comment (Borba02) [reply Tabloid/client/src/components/posts/PostDetails.js]: I have corrected that naming convention and gave some others more intention names.

File: Tabloid/client/src/components/ApplicationViews.js (hunk 3 above)
Comment (Borba02) [reply Tabloid/client/src/components/ApplicationViews.js]: They should. Added a check in ApplicationViews to address this

File: Tabloid/client/src/components/ApplicationViews.js [hunk 7]
```diff
@@ -20,10 +20,10 @@ export default function ApplicationViews({ isLoggedIn }) {
           <CategoryList />
         </Route>
         <Route path="/posts" exact>
-          <PostList />
+          {isLoggedIn ? <PostList /> : <Redirect to="/login" />}
         </Route>
-        <Route path="/posts/:postId" exact>
-          <PostDetails />
+        <Route path="/posts/:postId">
```
Comment (heymonicakay) [inline Tabloid/client/src/components/ApplicationViews.js]:
```suggestion
        <Route path="/posts/:postId(\d+)">
```

**Review by heymonicakay — CHANGES_REQUESTED** [review]
See my suggestion - DISREGARD

File: Tabloid/client/src/components/ApplicationViews.js (hunk 4 above)
Comment (Borba02) [reply Tabloid/client/src/components/ApplicationViews.js]: It doesn't, I've removed exact and added the regex needed to ensure the route param is an integer

File: Tabloid/Repositories/PostRepository.cs (hunk 1 above)
Comment (Borba02) [reply Tabloid/Repositories/PostRepository.cs]: I have added a new method to handle bools DbUtils.GetBool() and replaced them in PostRepository.cs

**Review by heymonicakay — APPROVED** [review]

**Pull request merged**